fvwm/fvwm.1
.\" Formatting instructions for the fvwm man page:
.\"
.\"  - Do not use \f... formatting instructions
.\"    unless you have to, see below.
.\"  - Avoid single and double quotes wherever possible.
.\"
.\" For further details, please refer to the Linux Man-Page how-to.
.\"
.\"  context                            typeface        example
.\"  ---------------------------------- --------------  -----------------
.\"  filenames:                         italics (.I)    .I .fvwm2rc
.\"  command line options of fvwm cmd:  bold (.B)       .B \-cmd
.\"  arguments of command line options: italics (.I)    .BI "\-cmd " command
.\"  built in commands:                 bold (.B)       .B Move
.\"  references to built in commands:   bold (.B)       .RB "See " Move
.\"  references to man page sections:   bold (.B)       .RB "See " SYNOPSIS
.\"  references to style options:       italics (.I)    .RI "See " Lenience
.\"  built in command options:          italics (.I)    .BI "Move " "0 0"
.\"  command syntax:                    bold/ita. (.BI) .BI "Move [" "x y" "]"
.\"  environment variables:             italics (.I)    .I $DISPLAY
.\"  key names:                         small (.SM)     .SM Tab
.\"  style names and strings:           double quotes   "stylename", "x"
.\"  single characters:                 single quotes   '@'
.\"  module names:                      bold (.B)       .B FvwmPager
.\"  links to web pages:                bold (.B)       .B http://fvwm.org
.\"  acronyms:                          small (.SM)     .SM ICCCM
.\"
.\"
.\" The name fvwm is written in lower case throughout this man
.\" page.  Fvwm is written without a version number, i.e. fvwm,
.\" not fvwm2.  The latter is used only in the .fvwm2rc file name.
.\"
.\" Note that the word "will" is rarely correct in a man page or
.\" any document. Describe what fvwm does, not what it will do,
.\" even if you haven't written the feature yet. dje 2/3/01
.\" (3/Feb/2001).
.\"
.\" A note to everyone who needs to write dates.  Don't use 2/3/01
.\" notation, a non-american person simply can't parse such
.\" sequence of digits.  Please use 3-Feb-2001 or 2001-02-03 forms.
.\" migo 16/May/2001.
.\"
.\" A note about the rule against \f... formatting.  On Solaris,
.\" AIX, and HPUX the ".BI" type commands are limited to 6
.\" arguments. use \f... formatting in that case.  See the lines
.\" formatted with ".IP".
.\"  dje 15/June/2001.
.\"
.\" @(#)fvwm-2.5.2 24 Jun 2002
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.ta .3i .6i .9i 1.2i 1.5i 1.8i
.TH FVWM 1 "24 June 2002" FVWM "FVWM 2.5.2"
.UC
.SH NAME

fvwm \- F(?) Virtual Window Manager (version 2) for X11
.SH SYNOPSIS

.B fvwm
.RB [ \-blackout ]
.RB [ \-clientId
.IR       id ]
.RB [ \-cmd
.IR       config_command ]
.RB [ \-d
.IR       displayname ]
.RB [ \-debug ]
.RB [ \-debug_stack_ring ]
.RB [ \-f
.IR       config_file ]
.RB [ \-h ]
.RB [ \-replace ]
.RB [ \-restore
.IR       state_file ]
.RB [ \-s ]
.RB [ \-version ]
.RB [ \-visual
.IR       visual_class ]
.RB [ \-visualId
.IR       id ]

.SH DESCRIPTION

Fvwm is a window manager for X11.  It is designed to minimize
memory consumption, provide a 3D look to window frames, and a
virtual desktop.

Note that there are several window managers around that have
"fvwm" in their name.  In the past, version 2.x of fvwm was
commonly called fvwm2 to distinguish it from the former version
1.x (fvwm or even fvwm1).  Since version 1.x has been replaced by
version 2.x a long time ago we simply call version 2.x and all
versions to come, fvwm, throughout this document, and the
executable program is named fvwm.  There is an fvwm offspring
called fvwm95, it is mostly a patch for fvwm-2.0.43.  The main
goal of fvwm95 was to supply a Windows 95 like look and feel.
Since then fvwm has been greatly enhanced and practically all
fvwm95 features can be imitated by fvwm.

Fvwm provides both a large
.I virtual desktop
and
.I multiple disjoint desktops
which can be used separately or together.  The virtual desktop
allows you to pretend that your video screen is really quite
large, and you can scroll around within the desktop.  The multiple
disjoint desktops allow you to pretend that you really have
several screens to work at, but each screen is completely
unrelated to the others.

Fvwm provides
.I keyboard accelerators
which allow you to perform most window-manager functions,
including moving and resizing windows, and operating the menus,
using keyboard shortcuts.

Fvwm has also blurred the distinction between configuration
commands and built-in commands that most window-managers make.
Configuration commands typically set fonts, colors, menu contents,
key and mouse function bindings, while built-in commands typically
do things like raise and lower windows.  Fvwm makes no such
distinction, and allows, to the extent that is practical, anything
to be changed at any time.

Other noteworthy differences between Fvwm and other X11 window
managers are the introduction of the
.IR SloppyFocus " and " NeverFocus
focus methods.  Focus policy can be specified for individual
windows.  Windows using
.I SloppyFocus
acquire focus when the pointer moves into them and retain focus
until some other window acquires it.  Such windows do not lose
focus when the pointer moves into the root window.  The
.I NeverFocus
policy is provided for use with windows into which one never types
(e.g. xclock, oclock, xbiff, xeyes, tuxeyes) - for example, if a
SloppyFocus terminal window has focus, moving the cursor over a
.I NeverFocus
decoration window won't deprive the terminal of focus.

.SH OPTIONS

These are the command line options that are recognized by fvwm:
.TP
.BI "-blackout"
This option is provided for backward compatibility only.  Blacking
out the screen during startup is not necessary anymore.
.TP
.BI "-clientId " id
This option is used when fvwm is started by a session
manager. Should not be used by a user.

.TP
.BI "-cmd " config_command
Causes fvwm to use
.I config_command
instead of
.RB "'" "Read"
.IR ".fvwm2rc" "'"
as its initialization command.  (Note that up to 10
.BR \-f " and " \-cmd
parameters can be given, and they are executed in the order
specified.)
.TP
.BI "-d " displayname
Manage the display called
.I displayname
instead of the name obtained from the environment variable
.IR "$DISPLAY" .
.TP
.BI "-debug"
Puts X transactions in synchronous mode, which dramatically slows
things down, but guarantees that fvwm's internal error messages
are correct. Also causes fvwm to output debug messages while
running.
.TP
.BI "-debug_stack_ring"
Enables stack ring debugging.  This option is only intended for
internal debugging and should only be used by developers.
.TP
.BI "-f " config_file
Causes fvwm to read
.I config_file
instead of
.I .fvwm2rc
as its initialization file.  This is equivalent to
.RB "-cmd 'Read"
.IR "config_file'."
.
.TP
.BI "-h"
A short usage description is printed.
.TP
.BI "-replace"
Try to take over from a previously running wm.  This does not work
unless the other wm is
.SM ICCCM
2.0 compliant.
.TP
.BI "-restore " state_file
This option is used when fvwm is started by a session manager.
Should not be used by a user.
.TP
.BI "-s"
On a multi-screen display, run fvwm only on the screen named in
the
.I $DISPLAY
environment variable or provided through the
.B -d
option. Normally, fvwm attempts to start up on all screens of a
multi-screen display.
.TP
.BI "-version"
Prints the version of fvwm to
.IR stderr .
Also prints an information about the compiled in support for
readline, rplay, stroke, xpm, png, gnome hints, EWMH hints,
session management, bidirectional text, multibyte characters,
xinerama and Xft aa font rendering.
.TP
.BI "-visual " visual_class
Causes fvwm to use
.I visual_class
for the window borders and menus.
.I visual_class
can be "StaticGray", "GrayScale", "StaticColor", "PseudoColor",
"TrueColor" or "DirectColor".
.TP
.BI "-visualId " id
Causes fvwm to use
.I id
as the visualId for the window borders and menus.
.I id
can be specified as N for decimal or 0xN for hexadecimal. See man
page of xdpyinfo for a list of supported visuals.


.SH ANATOMY OF A WINDOW

Fvwm puts a decorative border around most windows.  This border
consists of a bar on each side and a small L-shaped section on
each corner.  There is an additional top bar called the title-bar
which is used to display the name of the window.  In addition,
there are up to 10 title-bar buttons.  The top, side, and bottom
bars are collectively known as the side-bars.  The corner pieces
are called the frame.

Unless the standard defaults files are modified, pressing mouse
button 1 in the title or side-bars begins a move operation on the
window.  Pressing button 1 in the corner frame pieces begins a
resize operation.  Pressing button 2 anywhere in the border brings
up an extensive list of window operations.

Up to ten title-bar buttons may exist.  Their use is completely
user definable.  The default configuration has a title-bar button
on each side of the title-bar.  The one on the left is used to
bring up a list of window options, regardless of which mouse
button is used.  The one on the right is used to iconify the
window.  The number of title-bar buttons used depends on which
ones have mouse actions bound to them.  See the section on the
.B Mouse
command below.


.SH THE VIRTUAL DESKTOP

Fvwm provides multiple virtual desktops for users who wish to use
them.  The screen is a viewport onto a
.I desktop
which may be larger than the screen.  Several distinct desktops
can be accessed (concept: one desktop for each project, or one
desktop for each application, when view applications are
distinct).  Since each desktop can be larger than the physical
screen, divided into m by n
.I pages
which are each the size of the physical screen, windows which are
larger than the screen or large groups of related windows can
easily be viewed.

The (m by n) size (i.e. number of pages) of the virtual desktops
can be changed any time, by using the
.B DeskTopSize
built-in command.  All virtual desktops must be (are) the same
size.  The total number of distinct desktops does not need to be
specified, but is limited to approximately 4 billion total. All
windows on a range of desktops can be viewed in the
.BR FvwmPager ,
a miniature view of the desktops.  The pager is an accessory
program, called a module, which is not essential for the window
manager to operate.  Windows may also be listed, along with their
geometries, in a window list, accessible as a pop-up menu, or as a
separate window, called the
.B FvwmWinList
(another module).

Fvwm keeps the windows on the desktop in a layered stacking order;
a window in a lower layer never obscures a window in a higher
layer. The layer of a window can be changed by using the
.B Layer
command.  The concept of layers is a generalization of the
.I StaysOnTop
flag of older fvwm versions. The
.IR StaysOnTop " and " StaysPut
.B Style
options are now implemented by putting the windows in suitable
layers and the previously missing
.IB "StaysOnBottom " Style
option has been added.

.B Sticky
windows are windows which transcend the virtual desktop by
"Sticking to the screen's glass".  They always stay put on the
screen. This is convenient for things like clocks and xbiff's, so
you only need to run one such gadget and it always stays with you.

NEWS
Note, the changes for the last STABLE release start with release
2.6.0.

-------------------------------------------------------------------
Changes in stable release 2.6.8 (UNRELEASED)

* Bug fixes:

  - Various DESTDIR fixes (especially around the default-config
    Makefile)
  - fvwm-perllib pod2man fixes.
  - FvwmIconMan no longer triggers a warning about bad size hints
    in the fvwm core.
  - VMware windows do not disapper when switching between
    fullscreen and normal state.
  - Fix (de)installation with the configure options
    --program-prefix, --program-suffix and
    --program-transform-name.

-------------------------------------------------------------------
Changes in stable release 2.6.7 (06-Mar-2016)

* New fvwm features:

  - A new default configuration which is available when fvwm
    doesn't detect a configuration file to load.
  - A new conitional command "Desk n" can restrict matching
    windows to a specific desk.
  - A new conditional command "Screen n" to restrict matching
    windows on a given Xinerama screen.
  - New expansion variable "w.screen" to ascertain the Xineram
    screen number a window is on.
  - New command "InfoStoreClear" to remove all items in the
    InfoStore.

* Removed features:

  - The old and unmaintained debian/and rpm/ directories have
    been remmoved; use the maintainers' copies where available.
  - VMS support has been removed.
  - GTK1.x support has been removed.
  - GNOME-specific window hints (pre-EWMH) have been removed.
  - Some fvwm modules have been removed:
      - FvwmDragWell   (no replacement)
      - FvwmGTK        (no replacement)
      - FvwmSave       (no replacement)
      - FvwmSaveDesk   (no replacement)
      - FvwmScroll     (no replacement)
      - FvwmTabs       (no replacement, never worked anyway)
      - FvwmTaskBar    (use FvwmButtons)
      - FvwmTheme      (in core of fvwm as colorsets)
      - FvwmWharf      (use FvwmButtons)
      - FvwmWinList    (use WindowList command)
      - FvwmWindowMenu (use WindowList command)
      - FvwmIconBox    (use the IconBox style instead)

* New module features:

  - FvwmButtons learned a new option "Colorset" to its
    ChangeButton command.

* fvwm-menu-desktop updated:

  - Renamed default menu to XDGMenu and changed the name
    of the FvwmForm to FvwmForm-XDGMenu-Config to not conflict
    with someone already using FvwmMenu.
  - fvwm-menu-desktop will now load defaults from the
    FvwmForm-XDGMenu-Config data file.
  - Improved dynamic menus by regenerating them on-the-fly.
  - Added new options: --regen-cmd, --dynamic, and more.

* Bug fixes:

  - A bug introduced in 2.6.6 could cause applications with
    negative coordinates to be placed at strange positions.  This
    affected for example acroread when switching to fullscreen
    mode.  This has been fixed.
  - FvwmButtons "Silent" option for dynamic ChangeButton commands
    no longer loops infinitely.

Changes in stable release 2.6.6 (15-Mar-2016)

* New fvwm features:

  - Support for Russian from Ivan Gayevskiy.
  - EnvMatch supports infostore variables.
  - The option "forget" to the Maximize command allows to
    unmaximize a window without changing its size and position.
  - Windows shaded to a corner (NW, NE, SW, SE) are reduced to a
    small square.
  - New option "!raise" to the WarpToWindow command.
  - The new extended variables $[wa.x], $[wa.y], $[wa.width],
    $[wa.height] can be used to get the geometry of the EWMH
    working area, and $[dwa.x], $[dwa.y], $[dwa.width],
    $[dwa.height] can be used to get the geometry of the EWMH
    dynamic working area.
  - The Resize commands accept "wa" or "da" as a suffix of the
    width or height value.  If present, the value is a percentage
    of the width or height of the EWMH working area or the EWMH
    dynamic working area.
  - Fvwm is much more resilient against applications that flood the
    window manager with repeated events.

* New module features:

  - FvwmForm supports separator lines.
  - New FvwmIconMan options:
      IconAndSelectButton
      IconAndSelectColorset

* Bug fixes:

  - Provide a wrapper for the deprecation of XKeycodeToKeysym and
    use XkbKeycodeToKeysym() where appropriate.
  - fvwm-menu-desktop is re-written and provides better support of
    the XDG menu specification.
  - Fix fvwm-menu-desktop keyError bug.  Use "others" if no desktop
    environment found.
  - FvwmIconMan had problems displaying the hilight colour on some
    systems (64 bit issue?).
  - Globally active windows cannot take the focus if the style
    forbids programs to take focus themselves (style
    !FPFocusByProgram).
  - Windows no longer jump from one position to the other which
    could happen in some cases with SnapAttraction.  Windows now
    snap to the closest window (or screen edge).
  - Removing bindings had several strange side effects that are
    fixed now (removing too many bindings; old bindings showing up
    again after another is removed; possibly other effects).
  - Windows sometimes did not get expose events (i.e. did not
    redraw properly) if they were uncovered by moving a window
    above them.  This has been fixed.
  - FvwmConsole now causes much less network traffic.
  - Suppress bogus events sent to the modules when a window is
    resized with the mouse.
  - Properly handle the has_ref_window_moved flag for ResizeMove and
    ResizeMoveMaximize.
  - Removed some unnecessary redraws in FvwmPager.
  - The option "NoDeskLimitY" option of the GotoPage command did
    not work.
  - Negative coordinates in the "rectangle" option to the Menu
    commend did not work correctly.  This has been fixed.

extras/fvwmpython/Fvwm.py
#! /usr/bin/env python

"""
Copyright 1996, Barry A. Warsaw <bwarsaw@python.org>
Permission is granted to distribute this software freely
for any purpose as long as the copyright and this
copyright notice remain intact.


FVWM version 2 module interface framework for Python.
Requires Python 1.4 <http://www.python.org/>

Many thanks to Jonathan Kelley <jkelley@ugcs.caltech.edu> for his
original inspiration, fvwmmod.py which worked with FVWM version 1.

"""

__version__ = '1.0'

import struct
import os
import sys
import string
import regex
from types import *

_APPNAME = ''



# useful contants from fvwm/module.h -- not all are duplicated here

SynchronizationError = 'SynchronizationError'
PipeReadError = 'PipeReadError'

HEADER_SIZE = 4
DATASIZE = struct.calcsize('l')
START = 0xffffffff
WINDOW_NONE = 0

# convert from a C unsigned long to a Python long int
def ulong(x): return long(x) % 0x100000000L



Contexts = {'C_NO_CONTEXT': 0,
	    'C_WINDOW':     1,
	    'C_TITLE':      2,
	    'C_ICON':       4,
	    'C_ROOT':       8,
	    'C_FRAME':     16,
	    'C_SIDEBAR':   32,
	    'C_L1':        64,
	    'C_L2':       128,
	    'C_L3':       256,
	    'C_L4':       512,
	    'C_L5':      1024,
	    'C_R1':      2048,
	    'C_R2':      4096,
	    'C_R3':      8192,
	    'C_R4':     16384,
	    'C_R5':     32768,
	    }
for k, v in Contexts.items()[:]:
    Contexts[v] = k



Types = {'M_NEW_PAGE':            1,
	 'M_NEW_DESK':         1<<1,
	 'M_ADD_WINDOW':       1<<2,
	 'M_RAISE_WINDOW':     1<<3,
	 'M_LOWER_WINDOW':     1<<4,
	 'M_CONFIGURE_WINDOW': 1<<5,
	 'M_FOCUS_CHANGE':     1<<6,
	 'M_DESTROY_WINDOW':   1<<7,
	 'M_ICONIFY':          1<<8,
	 'M_DEICONIFY':        1<<9,
	 'M_WINDOW_NAME':      1<<10,
	 'M_ICON_NAME':        1<<11,
	 'M_RES_CLASS':        1<<12,
	 'M_RES_NAME':         1<<13,
	 'M_END_WINDOWLIST':   1<<14,
	 'M_ICON_LOCATION':    1<<15,
	 'M_MAP':              1<<16,
	 'M_ERROR':            1<<17,
	 'M_CONFIG_INFO':      1<<18,
	 'M_END_CONFIG_INFO':  1<<19,
	 'M_ICON_FILE':        1<<20,
	 'M_DEFAULTICON':      1<<21,
	 'M_STRING':           1<<22,
	 }
for k, v in Types.items():
    Types[v] = k



# Window flags.  See fvwm.h
Flags = {'STARTICONIC':       (1<<0),
	 'ONTOP':             (1<<1),	# does window stay on top
	 'STICKY':            (1<<2),	# Does window stick to glass?
	 'WINDOWLISTSKIP':    (1<<3),
	 'SUPPRESSICON':      (1<<4),
	 'NOICON_TITLE':      (1<<5),
	 'Lenience':          (1<<6),
	 'StickyIcon':        (1<<7),
	 'CirculateSkipIcon': (1<<8),
	 'CirculateSkip':     (1<<9),
	 'ClickToFocus':      (1<<10),
	 'SloppyFocus':       (1<<11),
	 'SHOW_ON_MAP':       (1<<12),
	 'BORDER':            (1<<13),	# Is this decorated with border
	 'TITLE':             (1<<14),	# Is this decorated with title
	 'MAPPED':            (1<<15),	# is it mapped?
	 'ICONIFIED':         (1<<16),  # is it an icon now?
	 'TRANSIENT':         (1<<17),	# is it a transient window?
	 'RAISED':            (1<<18),	# if a sticky window, needs raising?
	 'VISIBLE':           (1<<19),  # is the window fully visible
	 'ICON_OURS':         (1<<20),  # is the icon win supplied by the app?
	 'PIXMAP_OURS':       (1<<21),  # is the icon pixmap ours to free?
	 'SHAPED_ICON':       (1<<22),  # is the icon shaped?
	 'MAXIMIZED':         (1<<23),  # is the window maximized?
	 'DoesWmTakeFocus':   (1<<24),
	 'DoesWmDeleteWindow': (1<<25),
	 'ICON_MOVED':        (1<<26),  # has the icon been moved by the user?
	 'ICON_UNMAPPED':     (1<<27),	# was the icon unmapped, even though
					# the window is still
					# iconified (Transients)
	 'MAP_PENDING':       (1<<28),  # Sent an XMapWindow, but didn't
					# receive a MapNotify yet.
	 'HintOverride':      (1<<29),
	 'MWMButtons':        (1<<30),
	 'MWMBorders':        (1<<31),
	 }

for k, v in Flags.items():
    Flags[v] = k

ALL_COMMON_FLAGS = (Flags['STARTICONIC']    | Flags['ONTOP'] | 
		    Flags['STICKY']         | Flags['WINDOWLISTSKIP'] | 
		    Flags['SUPPRESSICON']   | Flags['NOICON_TITLE'] |
		    Flags['Lenience']       |
		    Flags['StickyIcon']     | Flags['CirculateSkipIcon'] |
		    Flags['CirculateSkip']  | Flags['ClickToFocus'] |
		    Flags['SloppyFocus']    | Flags['SHOW_ON_MAP'])



# Bit gravity.  See X11/X.h
Gravity = {'Forget':    0,
           'NorthWest': 1,
           'North':     2,
           'NorthEast': 3,
           'West':      4,
           'Center':    5,
           'East':      6,
           'SouthWest': 7,
           'South':     8,
           'SouthEast': 9,
           'Static':    10,
           }

for k, v in Gravity.items():
    Gravity[v] = k


class PacketReader:
    """Reads a packet from the given file-like object.

    Public functions: read_packet(fp)
    """
    def read_packet(self, fp):
	"""Reads a packet from `fp' a file-like object.

	Returns an instance of a class derived from Packet.  Raises
	PipeReadError if a valid packet could not be read.  Returns
	None if no corresponding Packet-derived class could be
	instantiated.

	"""
	data = fp.read(DATASIZE * 2)
	try:
	    start, typenum = struct.unpack('2l', data)
	    typenum = ulong(typenum)
	except struct.error:
	    raise PipeReadError
	if start <> START:
	    raise SynchronizationError
	try:
	    msgtype = Types[typenum]
	except KeyError:
	    print 'unknown FVWM message type:', typenum, '(ignored)'
	    return None
	classname = string.joinfields(
	    map(string.capitalize,
		string.splitfields(msgtype, '_')[1:]) + ['Packet'],
	    '')
	# create an instance of the class named by the packet type
	try:
	    return getattr(PacketReader, classname)(typenum, fp)
	except AttributeError:
	    raise RuntimeError, 'no packet class named: %s' % classname

    class _Packet:
	def __init__(self, typenum, fp):
	    # start of packet and type have already been read
	    data = fp.read(DATASIZE * 2)
	    self.pktlen, self.timestamp = struct.unpack('2l', data)
	    self.pktlen = ulong(self.pktlen)
	    self.timestamp = ulong(self.timestamp)
	    self.msgtype = typenum
	    self.pkthandler = string.joinfields(
		map(string.capitalize,
		    string.splitfields(Types[typenum], '_')[1:]),
		'')
    
	def _read(self, fp, *attrs):
	    attrcnt = len(attrs)
	    fmt = '%dl' % attrcnt
	    data = fp.read(DATASIZE * attrcnt)
	    attrvals = struct.unpack(fmt, data)
	    i = 0
	    for a in attrs:
		setattr(self, a, attrvals[i])
		i = i+1
    
    class NewPagePacket(_Packet):
	def __init__(self, type, fp):
	    PacketReader._Packet.__init__(self, type, fp)
	    self._read(fp, 'x', 'y', 'desk', 'max_x', 'max_y')
    
    class NewDeskPacket(_Packet):
	def __init__(self, type, fp):
	    PacketReader._Packet.__init__(self, type, fp)
	    self._read(fp, 'desk')
    
    class AddWindowPacket(_Packet):
	def __init__(self, type, fp):
	    PacketReader._Packet.__init__(self, type, fp)
	    self._read(fp, 'top_id', 'frame_id', 'db_entry', 'x', 'y',
		       'width', 'height', 'desk', 'flags',
		       'title_height', 'border_width',
		       'base_width', 'base_height',
		       'resize_width_incr', 'resize_height_incr',
		       'min_width', 'min_height',
		       'max_width', 'max_height',
		       'icon_label_id', 'icon_pixmap_id',
		       'gravity',
		       'text_color', 'border_color')
    
    class ConfigureWindowPacket(AddWindowPacket):
	pass
    
    class LowerWindowPacket(_Packet):
	def __init__(self, type, fp):
	    PacketReader._Packet.__init__(self, type, fp)
	    self._read(fp, 'top_id', 'frame_id', 'db_entry')
    
    class RaiseWindowPacket(LowerWindowPacket):
	pass
    
    class DestroyWindowPacket(LowerWindowPacket):
	pass
    
    class FocusChangePacket(LowerWindowPacket):
	def __init__(self, type, fp):
	    PacketReader.LowerWindowPacket.__init__(self, type, fp)
	    self._read(fp, 'text_color', 'border_color')
    
    class IconifyPacket(LowerWindowPacket):
	def __init__(self, type, fp):
	    PacketReader.LowerWindowPacket.__init__(self, type, fp)
	    self._read(fp, 'x', 'y', 'width', 'height')
    
    class IconLocationPacket(IconifyPacket):
	pass
    
    class DeiconifyPacket(LowerWindowPacket):
	pass
    
    class MapPacket(LowerWindowPacket):
	pass
    
    class _VariableLenStringPacket:
	"""Can only be used for multiple inheritance with Packet"""
	def __init__(self, fp, extraskip=3):
	    fieldlen = self.pktlen - HEADER_SIZE - extraskip
	    data = fp.read(DATASIZE * fieldlen)
	    index = string.find(data, '\000')
	    if index >= 0:
		data = data[:index]
	    self.data = string.strip(data)
    
    class WindowNamePacket(LowerWindowPacket, _VariableLenStringPacket):
	def __init__(self, type, fp):
	    PacketReader.LowerWindowPacket.__init__(self, type, fp)
	    PacketReader._VariableLenStringPacket.__init__(self, fp)
	    self.name = self.data
    
    class IconNamePacket(WindowNamePacket):
	pass
    
    class ResClassPacket(WindowNamePacket):
	pass
    
    class ResNamePacket(WindowNamePacket):
	pass
    
    class EndWindowlistPacket(_Packet):
	pass
    
    class StringPacket(_Packet, _VariableLenStringPacket):
	def __init__(self, type, fp):
	    PacketReader._Packet.__init__(self, type, fp)
	    dummy = fp.read(DATASIZE * 3)       # three zero fields
	    PacketReader._VariableLenStringPacket.__init__(self, fp)

    class ErrorPacket(StringPacket):
	def __init__(self, type, fp):
	    PacketReader.StringPacket.__init__(self, type, fp)
	    self.errmsg = self.data
    
    class ConfigInfoPacket(StringPacket):
	pass
    
    class EndConfigInfoPacket(_Packet):
	pass

    class IconFilePacket(WindowNamePacket):
	pass

    class DefaulticonPacket(StringPacket):
	pass



class FvwmModule:
    __usetk = None
    __done = 0

    def __init__(self, argv):
	if len(argv) < 6:
	    raise UsageError, \
		  'Usage: %s ofd ifd configfile appcontext wincontext' % \
		  argv[0]
	try:
	    self.__tofvwm   = os.fdopen(string.atoi(argv[1]), 'wb', 0)
	    self.__fromfvwm = os.fdopen(string.atoi(argv[2]), 'rb', 0)
	except ValueError:
	    raise UsageError, \
		  'Usage: %s ofd ifd configfile appcontext wincontext' % \
		  argv[0]
	global _APPNAME
	self.program = _APPNAME = argv[0]
	self.configfile = argv[3]
	self.appcontext = argv[4]
	self.wincontext = argv[5]
	self.args = argv[6:][:]
	self.__pktreader = PacketReader()
	# initialize dispatch table
	self.__dispatch = {}

    def __close(self):
	self.__tofvwm.close()
	self.__fromfvwm.close()

    def __del__(self):
	self.__close()

    def __do_dispatch(self, *args):
	"""Dispatch on the read packet type.

	Any register()'d callbacks take precedence over instance methods.

	"""
	try:
	    pkt = self.__pktreader.read_packet(self.__fromfvwm)
	except PipeReadError:
	    self.__close()
	    sys.exit(1)
	# dispatch
	if pkt:
## 	    debug('dispatching: %s' % pkt.pkthandler)
	    if self.__dispatch.has_key(pkt.pkthandler):
		self.__dispatch[pkt.pkthandler](self, pkt)
	    elif hasattr(self, pkt.pkthandler):
		method = getattr(self, pkt.pkthandler)
		method(pkt)
	    else:
		self.unhandled_packet(pkt)

    #
    # alternative callback mechanism (used before method invocation)
    #
    def register(self, pktname, callback):
	"""Register a callback for a packet type.

	PKTNAME is the name of the packet to dispatch on.  Packet
	names are strings as described in the Fvwm Module Interface
	documentation.  PKTNAME can also be the numeric equivalent of
	the packet type.

	CALLBACK is a method that takes two arguments.  The first
	argument is the FvwmModule instance receiving the packet, and
	the second argument is the packet instance.

	Callbacks are not nestable.  If a callback has been previously
	registered for the packet, the old callback is returned.

	"""

	if type(pktname) == IntType:
	    pktname = Types[pktname]
	# raise KeyError if it is an invalid packet name
	Types[pktname]
	pkthandler = string.joinfields(
	    map(string.capitalize,
		string.splitfields(pktname, '_')[1:]),
	    '')
	try:
	    cb = self.__dispatch[pkthandler]
	except KeyError:
	    cb = None
	self.__dispatch[pkthandler] = callback
	return cb

    def unregister(self, pktname):
	"""Unregister any callbacks for the named packet.

	Any registered callback is returned.

	"""
	if type(pktname) == IntType:
	    pktname = Types[pktname]
	# raise KeyError if it is an invalid packet name
	Types[pktname]
	pkthandler = string.joinfields(
	    map(string.capitalize,
		string.splitfields(pktname, '_')[1:]),
	    '')
	try:
	    cb = self.__dispatch[pkthandler]
	    del self.__dispatch[pkthandler]
	except KeyError:
	    cb = None
	return cb

    #
    # starting and stopping the main loop
    #
    def start(self):
	# use a Tkinter mainloop only if Tkinter has been imported and
	# a Tk widget has been created.  Poking on
	# Tkinter._default_root is a kludge.
	if not self.__usetk:
	    try:
		t = sys.modules['Tkinter']
		self.__usetk = t._default_root
		if self.__usetk:
		    t.tkinter.createfilehandler(
			self.__fromfvwm,
			t.tkinter.READABLE,
			self.__do_dispatch)
	    except KeyError:
		pass
	if self.__usetk:
	    self.__usetk.mainloop()
	else:
	    self.__done = 0
	    while not self.__done:
		self.__do_dispatch()

    def stop(self):
	if self.__usetk:
	    self.__usetk.quit()
	else:
	    self.__done = 1

    #
    # misc
    #

    def send(self, command, window=WINDOW_NONE, cont=1):
	if command:
	    self.__tofvwm.write(struct.pack('l', window))
	    self.__tofvwm.write(struct.pack('i', len(command)))
	    self.__tofvwm.write(command)
	    self.__tofvwm.write(struct.pack('i', cont))
	    self.__tofvwm.flush()

    def set_mask(self, mask=None):
	if mask is None:
	    mask = 0
	    for name, flag in Types.items():
		if type(name) <> StringType:
		    continue
		methodname = string.joinfields(
		    map(string.capitalize,
			string.splitfields(name, '_')[1:]),
		    '')
		if hasattr(self, methodname) or \
		   self.__dispatch.has_key(methodname):
		    mask = mask | flag
	self.send('Set_Mask ' + `mask`)

    def unhandled_packet(self, pkt):
	debug('Unhandled packet type: %s' % Types[pkt.msgtype])

    #
    # useful shortcuts
    #
    def __override(self, callback):
	"""Returns a `callback cache' object for __restore()"""
	cache = {}
	for pktname in Types.keys():
	    if type(pktname) == StringType:
		cache[pktname] = self.register(pktname, callback)
	return cache

    def __restore(self, cache):
	for pktname, callback in cache.items():
	    if callback:
		self.register(pktname, callback)
	    else:
		self.unregister(pktname)

    def get_configinfo(self):
	"""Returns Fvwm module configuration line object."""
	info = []
	def callback(self, pkt, info=info):
	    pktname = Types[pkt.msgtype]
	    if pktname == 'M_CONFIG_INFO':
		info.append(string.strip(pkt.data))
	    elif pktname == 'M_END_CONFIG_INFO':
		self.stop()
	    else:
		error('Unexpected packet type: %s' % pktname)

	cache = self.__override(callback)
	self.send('Send_ConfigInfo')
	self.start()
	self.__restore(cache)
	return ConfigInfo(info)

    def get_windowlist(self):
	"""Returns an Fvwm window list dump as an instance."""
	import Tracker
	t = Tracker.Tracker()
	def callback(self, pkt, tracker=t):
	    if Types[pkt.msgtype] == 'M_END_WINDOWLIST':
		self.stop()
	    else:
		tracker.feed_pkt(pkt)

	cache = self.__override(callback)
	self.send('Send_WindowList')
	self.start()
	self.__restore(cache)
	return t



class ConfigInfo:
    """Class encapsulating the FVWM configuration lines.

    Public instance variables:

        iconpath   -- the value of IconPath variable
	pixmappath -- the value of PixmapPath variable
	clicktime  -- the value of ClickTime variable

    Public methods:

	get_infolines(re) -- returns all lines.  If optional RE is given, only
	                     lines that start with `*<RE>' are
	                     returned (the initial star should not be
	                     included).

    """

    def __init__(self, lines):
	self.__lines = lines[:]
	self.iconpath = self.__get_predefineds('iconpath')
	self.pixmappath = self.__get_predefineds('pixmappath')
	self.clicktime = self.__get_predefineds('clicktime')

    def __get_predefineds(self, varname):
	for line in self.__lines:
	    parts = string.split(line)
	    if string.lower(parts[0]) == varname:
		return parts[1]

    def get_infolines(self, re=None):
	pred = None
	if re:
	    cre = regex.compile('\*' + re)
	    pred = lambda l, cre=cre: cre.match(l) >= 0
	return filter(pred, self.__lines)



def error(msg, exit=None, flag='ERROR'):
    sys.stderr.write('[%s] <<%s>> %s\n' % (_APPNAME, flag, msg))
    if exit is not None:
	sys.exit(exit)

def debug(msg):
    error(msg, flag='DEBUG')



# testing

if __name__ == '__main__':
    def printlist(list):
	for item in list:
	    print "`%s'" % item

    m = FvwmModule(sys.argv)
    info = m.get_configinfo()
    print 'IconPath   ==', info.iconpath
    print 'PixmapPath ==', info.pixmappath
    print 'ClickTime  ==', info.clicktime
    print '1 =========='
    printlist(info.get_infolines())
    print '2 =========='
    printlist(info.get_infolines('FvwmIdent'))
